modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -138,6 +152,177 @@ public IgniteRel visit(IgniteTableScan rel) {
         return rel;
     }
 
+    private static class ModifyNodeShuttle extends IgniteRelShuttle {
+        List<RexNode> projections = null;
+        List<List<RexNode>> expressions = null;
+        boolean unionRaised = false;
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteRel visit(IgniteProject rel) {
+            if (!unionRaised) {
+                // projection before union
+                assert projections == null;
+                // unexpected branch
+                if (projections != null) {
+                    throw Util.FoundOne.NULL;
+                }
+                projections = rel.getProjects();
+            } else {
+                // projections after union
+                if (expressions == null) {
+                    expressions = new ArrayList<>();
+                }
+                expressions.add(rel.getProjects());
```

Comment:
what if plan will look like this 
```
UnionAll
  Project
    Project
      Values
    
```
?

Reply:
i append 1 level resolver, i think it would be enough, if (for some reasons) no - seems we need to implement logic for input resolver like used in rules.